lib/fog/compute/google/requests/reset_windows_password.rb

```diff
@@ -0,0 +1,140 @@
+
+          # Create new metadata object
+          new_metadata = update_windows_keys(old_metadata, metadata_entry)
+
+          # Set metadata on instance
+          server.set_metadata(new_metadata, false)
+
+          # Get encrypted password from Serial Port 4 Output
+          # TODO: Should use server.serial_port_output
+          # but unable to get port 4 right now
+
+          # If machine is booting for the first time, there appears to be a
+          # delay before the password appears on the serial port.
+          serial_port_output = get_server_serial_port_output(instance, zone, :port => 4).contents
+          while serial_port_output.empty?
+            sleep(5)
+            serial_port_output = get_server_serial_port_output(instance, zone, :port => 4).contents
+          end
+
+          # Parse and decrypt password
+          enc_password = get_encrypted_password_from_serial_port(serial_port_output, modulus)
+          password = decrypt_password(enc_password, key)
+
+          return password
+        end
+
+        def get_modulus_exponent_in_base64(key)
+          mod = [key.n.to_s(16)].pack("H*").strip
+          exp = [key.e.to_s(16)].pack("H*").strip
+          modulus = Base64.strict_encode64(mod).strip
+          exponent = Base64.strict_encode64(exp).strip
+          return modulus, exponent
+        end
+
+        def get_expiration_time_string
+          utc_now = Time.now.utc
+          expire_time = utc_now + 5 * 60
+          return expire_time.strftime("%Y-%m-%dT%H:%M:%SZ")
+        end
+
+        def get_json_string(user, modulus, exponent, email)
+          expire = get_expiration_time_string
+          data = { 'userName': user,
+                   'modulus': modulus,
+                   'exponent': exponent,
+                   'email': email,
+                   'expireOn': expire }
+          return ::JSON.dump(data)
+        end
+
+        def update_windows_keys(old_metadata, metadata_entry)
+          if old_metadata[:items]
+            new_metadata = Hash[old_metadata[:items].map { |item| [item[:key], item[:value]] }]
+          else
+            new_metadata = {}
+          end
+          new_metadata["windows-keys"] = metadata_entry
+          return new_metadata
+        end
+
+        def get_encrypted_password_from_serial_port(serial_port_output, modulus)
+          output = serial_port_output.split("\n")
+          output.reverse.each do |line|
+            begin
+              entry = ::JSON.parse(line)
+              if modulus == entry["modulus"]
+                return entry["encryptedPassword"]
+              end
+            rescue ::JSON::ParserError
+              next
+            end
+          end
+        end
+
+        def decrypt_password(enc_password, key)
+          decoded_password = Base64.strict_decode64(enc_password)
+          return key.private_decrypt(decoded_password, OpenSSL::PKey::RSA::PKCS1_OAEP_PADDING)
```

Comment:
Nit: OpenSSL can be tricky (esp. between platforms), so just to check - do you know if there can there be a situation here where this decrypt would fail? E.g. padding check errors. Or is the key always set by us?

If not - feel free to ignore.

If yes - please consider adding a relevant rescue here with a human-readable error message as OpenSSL errors can be somewhat cryptic and as such - hard to debug without good knowledge of code that threw them.

Reply:
There should not be an error here, however I did receive a strange error here the other day multiple times.  A padding check error appeared and I could not fix it, then it disappeared after waiting awhile.

We create the key, send the public portion to Google to use, and then decrypt what is returned.  We even use the modulus included with the encrypted password to confirm the correct key was used.

The only reason I can think this would happen would be if for some reason the value returned from Google or the metadata values we set has been corrupted.

I've added a rescue with this information to at least point someone in the right direction if this ever were to happen again.

Reply:
SGTM. Thank you!